### PR TITLE
community/php7: upgrade to 7.2.4

### PR DESCRIPTION
**Pull request by TBK**
This PR does
- upgrades PHP next version which has backward incompatible changes
- creates `community/php7-mcrypt`  as this extension no longer shipped
- removes `testing/php7-libsodium` as now it shipped with main package
- moves `testing/php7-cassandra` to unmaintained because it has no PHP 7.2 compatibility
- bumps all php package versions

### CODE REVIEW

**Comment by andypost** [comment]
That looks good except
- why asm patch removed? Iirr it was just partially fixed in upstream
- a lot of packages could be broken cos 7.2 has serious deprecations, so maybe start with testing as we used to todo in 7.0->7.1 transition?

**Comment by andypost** [comment]
Mcrypt & backward compatibility are my primary concerns http://php.net/manual/en/migration72.incompatible.php

**Comment by adduc** [comment]
While mcrypt is no longer provided with PHP, it is available through [PECL](https://pecl.php.net/package/mcrypt) and could be provided with a separate package (as many PECL packages are provided already).

**Comment by andypost** [comment]
Yes, that's what I mean  - this migration should be planned. Every php7 dependency needs analisys about compatibility

**Comment by TBK** [comment]
@andypost Your patch and commit to php-src looks identical to me - https://github.com/php/php-src/commit/8c8679207ae5ada8751288e75b78c928a4d34d1d#diff-d7e2676e71a32db62437624b49497d02

I will add a package for mcrypt.

**Comment by TBK** [comment]
I do not what the Edge's if backwards compatability policy allowes stuff to be broken or if it should be avoided.

File: testing/php7-libsodium/APKBUILD [hunk 1]
```diff
@@ -1,31 +0,0 @@
-# Contributor: Fabio Ribeiro <fabiorphp@gmail.com>
-# Maintainer: Fabio Ribeiro <fabiorphp@gmail.com>
-pkgname=php7-libsodium
```
Comment (andypost) [inline testing/php7-libsodium/APKBUILD]: this removal should be replaced with php7-mcrypt into community repo

**Comment by TBK** [comment]
That is my plan, will have some time tomorrow.

**Comment by jirutka** [comment]
> maybe start with testing as we used to todo in 7.0->7.1 transition?

You can create a temporary aport `php72` in testing, but do not duplicate `php7` in testing, this caused many problems last time. However, it'd be better if we can avoid it and update all php stuff at once.

**Comment by andypost** [comment]
After checking other extensions I suggest to create `php72` cos every extension needs to be checked against new api, and least main debugging tool https://bugs.xdebug.org/roadmap_page.php?version_id=51 still not complete

**Comment by andypost** [comment]
I did check of php extensions and filed few PRs to fix, mostly all others' except xdebug (which could have special "alpha" version in testing) should work with php72

So the only remaining troubles are packages that depend on php7 - I found only `phpldapadmin` that depends on php5 https://github.com/leenooks/phpLDAPadmin/issues/43

**Comment by jirutka** [comment]
@andypost What is the current state, can we already upgrade php7 to 7.2?

File: community/php7/APKBUILD [hunk 2]
```diff
@@ -326,7 +324,7 @@ _build() {
 		--enable-zip=shared \
 			--with-libzip=/usr \
 		--with-zlib=shared \
-		--with-zlib-dir=/usr \
+			--with-zlib-dir=/usr \
```
Comment (andypost) [inline community/php7/APKBUILD]: Do we need this line after https://github.com/alpinelinux/aports/pull/2951/files

**Comment by andypost** [comment]
That looks good to go, zlib could be fixed in follow-up if needed

**Comment by andypost** [comment]
Fresher security release 7.2.1 is out http://php.net/archive/2018.php#id2018-01-04-2

**Comment by TBK** [comment]
Bumped to 7.2.1.

**Comment by andypost** [comment]
xdebug released for 7.2 here upgrade https://github.com/alpinelinux/aports/pull/3177

**Comment by TBK** [comment]
Great :)

So after #3177 this PR should be ready to go.

**Comment by ncopa** [comment]
I think it needs to be rebased

**Comment by TBK** [comment]
@ncopa Done.

**Comment by andypost** [comment]
7.2.2 is out http://php.net/archive/2018.php#id2018-02-01-1

I also filed https://github.com/alpinelinux/aports/pull/3222

**Comment by TBK** [comment]
@andypost @ncopa @jirutka 7.2.2 ready to go 👍

**Comment by andypost** [comment]
The last question for me is how we will maintain 7.1.x in 3.7 release only security releases will be backported?

**Comment by andypost** [comment]
It needs rebase on top https://github.com/alpinelinux/aports/pull/3222

**Comment by jirutka** [comment]
> The last question for me is how we will maintain 7.1.x in 3.7 release only security releases will be backported?

Yes. And we should finally upgrade php7 to 7.2 in edge.

**Comment by TBK** [comment]
@andypost @jirutka Rebase complete.

**Comment by andypost** [comment]
Looks everything ready for the change, I expect reports about incompatibilities because lots of apps is not adopted for 7.x to run (for example Drupal scheduled to start of March to release 7.2 compatible version
I checked `drupal7` package and it is sticked on `php5` so is safe
The rest of extensions reported compatible with 7.2

Guess it's time to decide when the switch will be made to get ready for feedback

**Comment by andypost** [comment]
I filed new aport for profiler that compatible with 7.1 & 7.2 https://github.com/alpinelinux/aports/pull/3303

**Comment by andypost** [comment]
@TBK please update PR with php 7.2.3 I think it is time to get this in

**Comment by emirb** [comment]
Can this get out please?

**Comment by TBK** [comment]
I have bumped all extensions except for one testing\php7-cassandra I can not get it to compile. Plus done some modernizing (apkbuild format, styling, added tests, updating).

On of the checks fails and I can not figure out why, any assistance would be appreciated:
```
checking for cass_cluster_new in -lcassandra... no
configure: error: Unable to load libcassandra
```

**Comment by andypost** [comment]
@TBK `php7-cassadnra` has no compatible release yet, ref https://datastax-oss.atlassian.net/projects/PHP/issues/PHP-221 looks there's WIP so probably better to remove that extension and create new PR to make it back when it is ready

The same applies to `libcouchbase` and related - please decouple it with `perl` to separate PR

Lots of impressive clean-ups for all php extensions are impressive! they could go this PR

https://github.com/alpinelinux/aports/pull/3519 is about upgrade of gmagick, I could commit it asap

**Comment by TBK** [comment]
@andypost It was and still is my intention to split the couchbase commits out into its own PR. I first saw your gmagick PR after pushing xD

**Comment by andypost** [comment]
@TBK please review #3519 cos it is what it missing to be commited) and I see this way all php extensions should be made

**Comment by andypost** [comment]
btw I can't find the WIP branch for cassandra 7.2 compatibility(

**Comment by TBK** [comment]
I found the old repo (https://github.com/datastax/php-dse-driver), it seem that Datastax has removed the code only offer the DSE variant to customers.

**Comment by andypost** [comment]
Then my eyes are fine) Then better to remove the extension while it's not compatible

**Comment by TBK** [comment]
What about moving it to unmaintained?

**Comment by andypost** [comment]
@TBK so only version bump for `testing/php7-gmagick` is left here

**Comment by TBK** [comment]
@andypost Done 💃 👍

**Comment by andypost** [comment]
I'm ++ to get this in and https://github.com/alpinelinux/aports/pull/3716 as follow-up

File: community/php7-imagick/APKBUILD [hunk 3]
```diff
@@ -12,6 +12,8 @@ arch="all"
 license="PHP"
 depends="php$_phpver-common"
 makedepends="php$_phpver-dev autoconf libtool imagemagick-dev pcre-dev"
+checkdepends="imagemagick"
```
[the remark this answers is not in the document]
Comment (andypost) [reply community/php7-imagick/APKBUILD]: https://github.com/alpinelinux/aports/pull/3764

[the remark this answers is not in the document]
Comment (TBK) [reply community/php7-imagick/APKBUILD]: @andypost I will rebase and bump when #3764 gets merged.

[the remark this answers is not in the document]
Comment (andypost) [reply community/php7-imagick/APKBUILD]: Sure, fyi https://github.com/alpinelinux/aports/pull/3770

[the remark this answers is not in the document]
Comment (andypost) [reply community/php7-imagick/APKBUILD]: both changes approved by maintainer & merged!

**Comment by TBK** [comment]
@andypost Rebased and bumped gmagick, imagick and couchbase.

**Comment by andypost** [comment]
Looks I need to rebuild all packages locally and then try to install them to be sure everything updated and works.
@TBK did you test them all? probably some packages' improvements still better to file as separate PRs to have only version bump here, can you care about it?

**Comment by TBK** [comment]
@andypost I have only used the check tests, no manual stuff.

I have made a PR for the modernization stuff - https://github.com/alpinelinux/aports/pull/3784

File: community/php7-mailparse/APKBUILD [hunk 4]
```diff
@@ -24,6 +24,11 @@ build() {
 	make
 }
 
+check() {
+	cd "$builddir"
+	make NO_INTERACTION=1 REPORT_EXIT_STATUS=1 test
```
Comment (andypost) [inline community/php7-mailparse/APKBUILD]: this tests somehow broken
```
Warning: PHP Startup: Unable to load dynamic library '/mnt/community/php7-mailparse/src/mailparse-3.0.2/modules/mailparse.so' - Error relocating /mnt/community/php7-mailparse/src/mailparse-3.0.2/modules/mailparse.so: mbfl_convert_filter_new: symbol not found in Unknown on line 0
```

Comment (TBK) [reply community/php7-mailparse/APKBUILD]: It is not loading mbstring.so

File: community/php7-memcached/APKBUILD [hunk 5]
```diff
@@ -27,6 +27,11 @@ build() {
 	make
 }
 
+check() {
+	cd "$builddir"
```
Comment (andypost) [inline community/php7-memcached/APKBUILD]: no reason to add because check disabled & require memcache server

File: community/php7-redis/APKBUILD [hunk 6]
```diff
@@ -24,6 +24,11 @@ build() {
 	make
 }
 
+check() {
```
Comment (andypost) [inline community/php7-redis/APKBUILD]: the same, check is disabled

File: testing/php7-solr/APKBUILD [hunk 7]
```diff
@@ -24,6 +24,11 @@ build() {
 	make
 }
 
+check() {
+	cd "$builddir"
```
Comment (andypost) [inline testing/php7-solr/APKBUILD]: the same

File: testing/php7-mustache/APKBUILD [hunk 8]
```diff
@@ -39,7 +38,7 @@ build() {
 
 check() {
 	cd "$builddir"
-	make test < /dev/null
+	make NO_INTERACTION=1 REPORT_EXIT_STATUS=1 test
```
Comment (andypost) [inline testing/php7-mustache/APKBUILD]: makes sense to separate PR

File: testing/php7-msgpack/APKBUILD [hunk 9]
```diff
@@ -3,33 +3,34 @@
...
-	phpize7 || return 1
-	./configure --prefix=/usr --with-php-config=php-config7 || return 1
-	make || return 1
+	phpize7
+	./configure --prefix=/usr --with-php-config=php-config7
+	make
+}
+
+check() {
+	cd "$builddir"
```
Comment (andypost) [inline testing/php7-msgpack/APKBUILD]: the same, no reason for check() & needs separate PR

**Comment by andypost** [comment]
So only cassandra fails, I'm sure it's totally fine for unmaintained
```
Successfully build packages: community/php7 community/php7-amqp community/php7-apcu community/php7-gmagick community/php7-imagick community/php7-mailparse community/php7-mcrypt community/php7-memcached community/php7-oauth community/php7-redis community/php7-ssh2 community/php7-xdebug community/php7-yaml community/php7-zmq testing/php7-ast testing/php7-couchbase testing/php7-inotify testing/php7-lzf testing/php7-mongodb testing/php7-msgpack testing/php7-mustache testing/php7-rdkafka testing/php7-solr testing/php7-tideways_xhprof testing/php7-timezonedb testing/php7-uuid
Failed to build packages: unmaintained/php7-cassandra
```

**Comment by andypost** [comment]
@TBK meanwhile 7.2.4 released http://php.net/archive/2018.php#id2018-03-29-2
Please update PR - I'm going to accept it tonight

**Comment by TBK** [comment]
@andypost Done. I do not want to put any more work into this PR...

**Comment by andypost** [comment]
Here's a list of test failures on 7.2.4 https://gist.github.com/andypost/25a1673263d09cacacd485cccf53676a

**Comment by TBK** [comment]
I can add them all to the disabled test list (https://git.alpinelinux.org/cgit/aports/tree/community/php7/disabled-tests.list) if there is not an obvious fix to resolve the issues.

**Comment by andypost** [comment]
@TBK I see no reason to reroll it anymore, let's get this in and fix broken tests after build in separate PRs

**Comment by TBK** [comment]
@andypost I just wanted to say thanks for all your work related to this PR 👍

**Comment by andypost** [comment]
@TBK thanks to you too, so many re-rolls and clean-ups!
This weekend I gonna debug tests, as expected the failures mostly different for each arch https://gist.github.com/andypost/25a1673263d09cacacd485cccf53676a (added as comments from build system)

**Comment by f3l1x** [comment]
Thank you guys.

**Pull request closed without merging**